File: benchmarks/nbody/nbody.lua
local PI = 3.141592653589793
local SOLAR_MASS = 4 * PI * PI
local DAYS_PER_YEAR = 365.24

local function advance(bodies, dt)
    local n = #bodies
    for i = 1, n do
        for j = i + 1, n do
            local dx = bodies[i].x - bodies[j].x
            local dy = bodies[i].y - bodies[j].y
            local dz = bodies[i].z - bodies[j].z

            local dSquared = dx*dx + dy*dy + dz*dz
            local distance = math.sqrt(dSquared)
            local mag = dt / (dSquared * distance)

            bodies[i].vx = bodies[i].vx - dx * bodies[j].mass * mag
            bodies[i].vy = bodies[i].vy - dy * bodies[j].mass * mag
            bodies[i].vz = bodies[i].vz - dz * bodies[j].mass * mag

            bodies[j].vx = bodies[j].vx + dx * bodies[i].mass * mag
            bodies[j].vy = bodies[j].vy + dy * bodies[i].mass * mag
            bodies[j].vz = bodies[j].vz + dz * bodies[i].mass * mag
        end
    end

    for i = 1, n do
        bodies[i].x = bodies[i].x + dt * bodies[i].vx
        bodies[i].y = bodies[i].y + dt * bodies[i].vy
        bodies[i].z = bodies[i].z + dt * bodies[i].vz
    end
end

local function energy(bodies)
    local e = 0.0
    local n = #bodies
    for i = 1, n do
        e = e + 0.5 * bodies[i].mass * (bodies[i].vx*bodies[i].vx + bodies[i].vy*bodies[i].vy + bodies[i].vz*bodies[i].vz)
        for j = i + 1, n do
            local dx = bodies[i].x - bodies[j].x
            local dy = bodies[i].y - bodies[j].y
            local dz = bodies[i].z - bodies[j].z
            local distance = math.sqrt(dx*dx + dy*dy + dz*dz)
            e = e - (bodies[i].mass * bodies[j].mass) / distance
        end
    end
    return e
end

local function offsetMomentum(bodies)
    local px, py, pz = 0.0, 0.0, 0.0
    for i = 1, #bodies do
        px = px + bodies[i].vx * bodies[i].mass
        py = py + bodies[i].vy * bodies[i].mass
        pz = pz + bodies[i].vz * bodies[i].mass
    end
    bodies[1].vx = -px / SOLAR_MASS
    bodies[1].vy = -py / SOLAR_MASS
    bodies[1].vz = -pz / SOLAR_MASS
end

local bodies = {
    -- Sun
    {x=0, y=0, z=0, vx=0, vy=0, vz=0, mass=SOLAR_MASS},
    -- Jupiter
    {x=4.84143144246472090e+00, y=-1.16032004402742839e+00, z=-1.03622044471123109e-01,
     vx=1.66007664274403694e-03 * DAYS_PER_YEAR, vy=7.69901118419740425e-03 * DAYS_PER_YEAR, vz=-6.90460016972063023e-05 * DAYS_PER_YEAR,
     mass=9.54791938424326609e-04 * SOLAR_MASS},
    -- Saturn
    {x=8.34336671824457987e+00, y=4.12479856412430479e+00, z=-4.03523417114321381e-01,
     vx=-2.76742510726862411e-03 * DAYS_PER_YEAR, vy=4.99852801234917238e-03 * DAYS_PER_YEAR, vz=2.30417297573763929e-05 * DAYS_PER_YEAR,
     mass=2.85885980666130812e-04 * SOLAR_MASS},
    -- Uranus
    {x=1.28943695621391310e+01, y=-1.51111514016986312e+01, z=-2.23307578892655734e-01,
     vx=2.96460137564761618e-03 * DAYS_PER_YEAR, vy=2.37847173959480950e-03 * DAYS_PER_YEAR, vz=-2.96589568540237556e-05 * DAYS_PER_YEAR,
     mass=4.36624404335156298e-05 * SOLAR_MASS},
    -- Neptune
    {x=1.53796971148509165e+01, y=-2.59193146099879641e+01, z=1.79258772950371181e-01,
     vx=2.68067772490389322e-03 * DAYS_PER_YEAR, vy=1.62824170038242295e-03 * DAYS_PER_YEAR, vz=-9.51592254519715870e-05 * DAYS_PER_YEAR,
     mass=5.15138902046611451e-05 * SOLAR_MASS},
}

offsetMomentum(bodies)
print(string.format("%.9f", energy(bodies)))

local n = 170000
for i = 1, n do
    advance(bodies, 0.01)
end

print(string.format("%.9f", energy(bodies)))

File: benchmarks/nbody/nbody.rb
PI = 3.141592653589793
SOLAR_MASS = 4 * PI * PI
DAYS_PER_YEAR = 365.24

class Body
  attr_accessor :x, :y, :z, :vx, :vy, :vz, :mass

  def initialize(x, y, z, vx, vy, vz, mass)
    @x = x
    @y = y
    @z = z
    @vx = vx
    @vy = vy
    @vz = vz
    @mass = mass
  end
end

def advance(bodies, dt)
  n = bodies.length
  (0...n).each do |i|
    ((i + 1)...n).each do |j|
      dx = bodies[i].x - bodies[j].x
      dy = bodies[i].y - bodies[j].y
      dz = bodies[i].z - bodies[j].z

      d_squared = dx*dx + dy*dy + dz*dz
      distance = Math.sqrt(d_squared)
      mag = dt / (d_squared * distance)

      bodies[i].vx -= dx * bodies[j].mass * mag
      bodies[i].vy -= dy * bodies[j].mass * mag
      bodies[i].vz -= dz * bodies[j].mass * mag

      bodies[j].vx += dx * bodies[i].mass * mag
      bodies[j].vy += dy * bodies[i].mass * mag
      bodies[j].vz += dz * bodies[i].mass * mag
    end
  end

  (0...n).each do |i|
    bodies[i].x += dt * bodies[i].vx
    bodies[i].y += dt * bodies[i].vy
    bodies[i].z += dt * bodies[i].vz
  end
end

def energy(bodies)
  e = 0.0
  n = bodies.length
  (0...n).each do |i|
    e += 0.5 * bodies[i].mass * (bodies[i].vx**2 + bodies[i].vy**2 + bodies[i].vz**2)
    ((i + 1)...n).each do |j|
      dx = bodies[i].x - bodies[j].x
      dy = bodies[i].y - bodies[j].y
      dz = bodies[i].z - bodies[j].z
      distance = Math.sqrt(dx*dx + dy*dy + dz*dz)
      e -= (bodies[i].mass * bodies[j].mass) / distance
    end
  end
  e
end

def offset_momentum(bodies)
  px = py = pz = 0.0
  bodies.each do |b|
    px += b.vx * b.mass
    py += b.vy * b.mass
    pz += b.vz * b.mass
  end
  bodies[0].vx = -px / SOLAR_MASS
  bodies[0].vy = -py / SOLAR_MASS
  bodies[0].vz = -pz / SOLAR_MASS
end

def main
  bodies = [
    # Sun
    Body.new(0, 0, 0, 0, 0, 0, SOLAR_MASS),
    # Jupiter
    Body.new(4.84143144246472090e+00, -1.16032004402742839e+00, -1.03622044471123109e-01,
             1.66007664274403694e-03 * DAYS_PER_YEAR, 7.69901118419740425e-03 * DAYS_PER_YEAR, -6.90460016972063023e-05 * DAYS_PER_YEAR,
             9.54791938424326609e-04 * SOLAR_MASS),
    # Saturn
    Body.new(8.34336671824457987e+00, 4.12479856412430479e+00, -4.03523417114321381e-01,
             -2.76742510726862411e-03 * DAYS_PER_YEAR, 4.99852801234917238e-03 * DAYS_PER_YEAR, 2.30417297573763929e-05 * DAYS_PER_YEAR,
             2.85885980666130812e-04 * SOLAR_MASS),
    # Uranus
    Body.new(1.28943695621391310e+01, -1.51111514016986312e+01, -2.23307578892655734e-01,
             2.96460137564761618e-03 * DAYS_PER_YEAR, 2.37847173959480950e-03 * DAYS_PER_YEAR, -2.96589568540237556e-05 * DAYS_PER_YEAR,
             4.36624404335156298e-05 * SOLAR_MASS),
    # Neptune
    Body.new(1.53796971148509165e+01, -2.59193146099879641e+01, 1.79258772950371181e-01,
             2.68067772490389322e-03 * DAYS_PER_YEAR, 1.62824170038242295e-03 * DAYS_PER_YEAR, -9.51592254519715870e-05 * DAYS_PER_YEAR,
             5.15138902046611451e-05 * SOLAR_MASS),
  ]

  offset_momentum(bodies)
  puts format("%.9f", energy(bodies))

  n = 170000
  n.times { advance(bodies, 0.01) }

  puts format("%.9f", energy(bodies))
end

main
